// vscode-markdown-languageservice/out/util/editBuilder.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceEditBuilder = void 0;
const lsp = require("vscode-languageserver-types");
class WorkspaceEditBuilder {
    constructor() {
        this.edit = {
            changes: {},
        };
    }
    replace(resource, range, newText) {
        const resourceKey = resource.toString();
        let edits = this.edit.changes[resourceKey];
        if (!edits) {
            edits = [];
            this.edit.changes[resourceKey] = edits;
        }
        edits.push(lsp.TextEdit.replace(range, newText));
    }
    getEdit() {
        return this.edit;
    }
    renameFile(targetUri, resolvedNewFilePath) {
        if (!this.edit.documentChanges) {
            this.edit.documentChanges = [];
        }
        this.edit.documentChanges.push(lsp.RenameFile.create(targetUri.toString(), resolvedNewFilePath.toString()));
    }
}
exports.WorkspaceEditBuilder = WorkspaceEditBuilder;
