// typescript-language-server/lib/hover.js
/*
 * Copyright (C) 2017, 2018 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
import { asDocumentation, asPlainText } from './protocol-translation.js';
export function asSignatureHelp(info) {
    return {
        activeSignature: info.selectedItemIndex,
        activeParameter: getActiveParameter(info),
        signatures: info.items.map(asSignatureInformation)
    };
}
function getActiveParameter(info) {
    const activeSignature = info.items[info.selectedItemIndex];
    if (activeSignature && activeSignature.isVariadic) {
        return Math.min(info.argumentIndex, activeSignature.parameters.length - 1);
    }
    return info.argumentIndex;
}
function asSignatureInformation(item) {
    const parameters = item.parameters.map(asParameterInformation);
    const signature = {
        label: asPlainText(item.prefixDisplayParts),
        documentation: asDocumentation({
            documentation: item.documentation,
            tags: item.tags.filter(x => x.name !== 'param')
        }),
        parameters
    };
    signature.label += parameters.map(parameter => parameter.label).join(asPlainText(item.separatorDisplayParts));
    signature.label += asPlainText(item.suffixDisplayParts);
    return signature;
}
function asParameterInformation(parameter) {
    return {
        label: asPlainText(parameter.displayParts),
        documentation: asDocumentation(parameter)
    };
}
